common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -34,6 +34,15 @@
  * Utility class for collation-aware UTF8String operations.
  */
 public class CollationAwareUTF8String {
+
+  public static boolean lowercaseMatchAt(final UTF8String l, final UTF8String r,
+      int pos, int len) {
+    if (len + pos > l.numChars() || pos < 0) {
+      return false;
+    }
+    return l.substring(pos, pos + len).toLowerCase().equals(r);
```

Comment:
Shouldn't this be?
```java
return l.substring(pos, pos + len).toLowerCase().equals(r.toLowerCase());
```

Reply:
Another problem with this is that we are trying to match a known number of characters `len`, which you want it to be `r.numChars()`. I don't think that the overall logic here works.
I believe what you want is something like (this the concept, I am sure it can be optimized):
```java
public static boolean lowercaseMatchAt(final UTF8String l, final UTF8String r, int pos) {
    assert(pos > 0);
    UTF8String lsubstr = l.substring(pos);
    UTF8String r1 = r.toLowerCase();
    for (int i = 1; i <= lsubstr.numChars(); ++i) {
       if (lsubstr.substring(0, i).toLowerCase().equals(r1)) return true;
    }
    return false;
}
```

Reply:
update: resolved in latest commit (we're no longer trying to match a known number of characters)

Reply:
> Shouldn't this be?
> 
> return l.substring(pos, pos + len).toLowerCase().equals(r.toLowerCase());

not here, because we know needle is not gonna change (we won't take any substrings of needle), then we can lowercase the needle only once (before the function call) in order to avoid multiple lowercasing (this is only important for `contains` which turns out to be quadratic - just like ICU StringSearch algorithm, while `startsWith` and `endsWith` are linear in nature because one bound is known in advance)